packages/CollapsibleChecklists/src/components/Group/Group.js

```diff
@@ -0,0 +1,120 @@
+import React from "react";
+import PropTypes from "prop-types";
+import Collapsible from "@paprika/collapsible";
+import textContent from "react-addons-text-content";
+import Item from "../Item";
+import groupStyles from "./Group.styles";
+
+const propTypes = {
+  children: PropTypes.oneOfType([PropTypes.node, PropTypes.arrayOf(PropTypes.node)]), // are probably an array of "Items", but could be a Spinner or anything else
+  isDisabled: PropTypes.bool,
+  onChange: PropTypes.func,
+  onExpand: PropTypes.func,
+  title: PropTypes.node.isRequired,
+};
+
+const defaultProps = {
+  children: [],
+  isDisabled: false,
+  onChange: () => {},
+  onExpand: () => {},
+};
+
+function useIsIndeterminate(checkboxRef) {
+  const [, setIsIndeterminate] = React.useState(null);
+  React.useEffect(() => {
+    setIsIndeterminate(checkboxRef.current.indeterminate);
+  }, [checkboxRef.current.indeterminate]);
+}
+
+function Group(props) {
+  const { children, isDisabled, onChange, onExpand, title } = props;
+  const [isCollapsed, setIsCollapsed] = React.useState(true);
+  const checkboxRef = React.useRef({});
+  useIsIndeterminate(checkboxRef);
+  let allAreChecked = React.Children.count(props.children) > 0;
+  let noneAreChecked = true;
+
+  React.Children.forEach(children, child => {
+    if (child.props.isChecked) {
+      noneAreChecked = false;
+    } else {
+      allAreChecked = false;
+    }
+  });
+
+  if (allAreChecked || noneAreChecked) {
+    checkboxRef.current.indeterminate = false;
+  } else if (!allAreChecked && !noneAreChecked) {
+    checkboxRef.current.indeterminate = true;
+  }
+
+  function toggleChildren() {
+    const childItems =
+      children.length > 0 ? children.filter(childItem => childItem.type.displayName === Item.displayName) : [];
+
+    if (childItems.length === 0) {
+      return;
+    }
+
+    const allChildItemsAreChecked =
+      childItems.filter(childItem => childItem.props.isChecked || childItem.props.isDisabled).length ===
+      childItems.length;
+
+    let childItemsToChange = [];
+    if (allChildItemsAreChecked) {
+      childItemsToChange = childItems.filter(childItem => !childItem.props.isDisabled);
+    } else {
+      childItemsToChange = childItems.filter(childItem => !childItem.props.isChecked && !childItem.props.isDisabled);
+    }
+
+    onChange(childItemsToChange);
```

Comment:
I'm interested in hearing others' feedback about this. While discussing the API for `<Sortable>` with @mikrotron I said that an `onChange` API based on React elements doesn't make a lot of sense to me. In this case, given a set of React elements, how would the consumer know what to do? They would need to somehow map those elements to their actual data to determine what is being checked/unchecked. The storybook story doesn't really handle this, so it's not obvious that there's an issue.

We might want to require some kind of `id` prop on the items to handle this. (can't use `key` because that wouldn't be visible to us)

Reply:
Since we expect an array, maybe just the `indexes` would be sufficient?

Reply:
The idea is that you always received the selectedIndexes as the first argument, and we could pass the full list of elements as second argument so you could do:

```js
function handleChange(indexes, list) {
  indexes.forEach(index => console.log(list[index].props.value))
}
```

Reply:
We only have a single `onChange` prop for the entire component now, so a single index would not be enough to describe a child checkbox, you'd also need an index of the group. (Alternatively, put a separate `onChange` prop on each group—which might be too repetitive though? Realistically a consumer is more likely to define a single callback handler for this component anyway I think?)

For the snippet above, seems like the list is redundant? The consumer is the one passing in those "values" in the first place, so they don't really need to be given them back. They already have some kind of state that they can iterate over.

I suggest creating a story that would better reflect the real world usage of this component (i.e. resembling the feature that will need it). That should make it easier to design.

Reply:
The consumer can compare on anything... I chose to compare by `key` (as I set key={team.name} but could have just as easily called it `foo={team.name}`. In fact i'll update the example to use `foobar`, and the comparison checks `props.foobar`

Reply:
I see, I'm -1 on this.